src/main/java/com/kodilla/ecommercee/service/ProductGroupService.java

```diff
@@ -0,0 +1,71 @@
+package com.kodilla.ecommercee.service;
+
+import com.kodilla.ecommercee.common.Validator;
+import com.kodilla.ecommercee.dto.ProductGroupDto;
+import com.kodilla.ecommercee.entity.Product;
+import com.kodilla.ecommercee.entity.ProductGroup;
+import com.kodilla.ecommercee.exception.EntityAlreadyExistsException;
+import com.kodilla.ecommercee.exception.EntityNotFoundException;
+import com.kodilla.ecommercee.mapper.ProductGroupMapper;
+import com.kodilla.ecommercee.repository.ProductGroupRepository;
+import com.kodilla.ecommercee.repository.ProductRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+
+@Service
+public class ProductGroupService {
+    private static final String NOT_FOUND = "ERROR: Group not found.";
+    private static final String GROUP_EXISTS = "ERROR: Group already exists.";
+
+    @Autowired
+    ProductGroupRepository groupRepository;
+
+    @Autowired
+    ProductRepository productRepository;
+
+    @Autowired
+    ProductGroupMapper groupMapper;
+
+    @Autowired
+    Validator validator;
+
+    public List<ProductGroupDto> getGroups() {
+        return groupMapper.mapToGroupDtoList(groupRepository.findAll());
+    }
+
+    public ProductGroupDto getGroup(Long id) {
+        ProductGroup group = groupRepository.findById(id).orElseThrow(() -> new EntityNotFoundException(NOT_FOUND));
+        return groupMapper.mapToGroupDto(group);
+    }
+
+    public void create(ProductGroupDto groupDto) {
+        validator.validateUser(groupDto.getKey());
+        if (groupRepository.existsByName(groupDto.getName())) throw new EntityAlreadyExistsException(GROUP_EXISTS);
+
+        groupRepository.save(groupMapper.mapToGroup(groupDto));
+    }
+
+    public ProductGroupDto update(ProductGroupDto groupDto) {
+        validator.validateUser(groupDto.getKey());
+        if (!groupRepository.existsById(groupDto.getId())) throw new EntityNotFoundException(NOT_FOUND);
+        ProductGroup group = groupMapper.mapToGroup(groupDto);
+
+        return groupMapper.mapToGroupDto(groupRepository.save(group));
+    }
+
+    public void delete(Long id, String key) {
+        validator.validateUser(key);
+        ProductGroup group = groupRepository.findById(id).orElseThrow(() -> new EntityNotFoundException(NOT_FOUND));
```

Comment:
Taki zapis usuwanie grupy jest mało czytelny. Jeśli chcemy faktycznie usuwać wszystkie produkty przypisane do grupy to możemy dodać cascade. Ewentualnie można by nie usuwać produktów, ale zrobić "grupę techniczną" typu "unassigned products" gdzie trafiają wszystkie nie przypisane produkty jeśli usunie się ich grupe.